src/webui/api/searchcontroller.cpp

```diff
@@ -0,0 +1,286 @@
+/*
+ * Bittorrent Client using Qt and libtorrent.
+ * Copyright (C) 2018  Thomas Piccirello <thomas.piccirello@gmail.com>
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * In addition, as a special exception, the copyright holders give permission to
+ * link this program with the OpenSSL project's "OpenSSL" library (or with
+ * modified versions of it that use the same license as the "OpenSSL" library),
+ * and distribute the linked executables. You must obey the GNU General Public
+ * License in all respects for all of the code used other than "OpenSSL".  If you
+ * modify file(s), you may extend this exception to your version of the file(s),
+ * but you are not obligated to do so. If you do not wish to do so, delete this
+ * exception statement from your version.
+ */
+
+#include "searchcontroller.h"
+
+#include <QJsonArray>
+#include <QJsonObject>
+
+#include "base/utils/misc.h"
+#include "base/utils/random.h"
+#include "base/utils/string.h"
+#include "apierror.h"
+
+class SearchPluginManager;
+
+using Utils::String::parseBool;
+
+void SearchController::startAction()
+{
+    checkParams({"pattern", "category", "plugins"});
+
+    if (Utils::Misc::pythonVersion() < 0)
+        throw APIError(APIErrorType::Conflict, "Python must be installed to use the Search Eengine.");
+
+    const QString pattern = params()["pattern"].trimmed();
+    const QString category = params()["category"].trimmed();
+    const QString plugins = params()["plugins"].trimmed();
+    const QString pluginsLower = plugins.toLower();
+
+    QStringList pluginsToUse;
+    if (pluginsLower == "all") pluginsToUse = SearchPluginManager::instance()->allPlugins();
+    else if (pluginsLower == "enabled") pluginsToUse = SearchPluginManager::instance()->enabledPlugins();
+    else if (pluginsLower == "multi") pluginsToUse = SearchPluginManager::instance()->enabledPlugins();
+    else pluginsToUse << plugins;
+
+    const QString id = generateSearchId();
+    auto *searchHandler = SearchPluginManager::instance()->startSearch(pattern, category, pluginsToUse);
+    m_searches[id] = new WebSearchHandler(searchHandler);
+
+    QVariantMap resultMap;
```

Comment:
```c++
const QJsonObject result = QJsonObject::fromVariantMap({ {"status", "Ok."}, {"id", id} });
```
should be possible...

Reply:
Why do you need this intermediate object? I mean QVariantMap.

Reply:
> should be possible...

I didn't know about this inline form. Thanks.

> Why do you need this intermediate object? I mean QVariantMap.

Because I don't know of another way. I'm still very new to C++, so when pointing out that there are other ways, suggestions are welcome.

Reply:
>suggestions are welcome

You should use QJson types directly, unless you already have data in different format.
QJsonObject has similar interface with QHash/QMap.
QJsonArray has similar interface with QList.

You can do either:
```c++
const QJsonObject result = {{"status", "Ok."}, {"id", id}};
setResult(result);
```
or (if you don't need intermediate variable):
```c++
setResult(QJsonObject {{"status", "Ok."}, {"id", id}});
```

So please don't create QVariantXXX types only to convert it to QJsonXXX types (everywhere in this code).